execution_engine/src/storage/global_state/scratch.rs

```diff
@@ -0,0 +1,495 @@
+    type Error = error::Error;
+
+    fn read(
+        &self,
+        correlation_id: CorrelationId,
+        key: &Key,
+    ) -> Result<Option<StoredValue>, Self::Error> {
+        if let Some(value) = self.cache.read().unwrap().get(key) {
+            return Ok(Some(value.clone()));
+        }
+        let txn = self.environment.create_read_txn()?;
+        let ret = match read::<Key, StoredValue, lmdb::RoTransaction, LmdbTrieStore, Self::Error>(
+            correlation_id,
+            &txn,
+            self.store.deref(),
+            &self.root_hash,
+            key,
+        )? {
+            ReadResult::Found(value) => {
+                self.cache.write().unwrap().insert(*key, value.clone());
+                Some(value)
+            }
+            ReadResult::NotFound => None,
+            ReadResult::RootNotFound => panic!("ScratchGlobalState has invalid root"),
+        };
+        txn.commit()?;
+        Ok(ret)
+    }
+
+    fn read_with_proof(
+        &self,
+        correlation_id: CorrelationId,
+        key: &Key,
+    ) -> Result<Option<TrieMerkleProof<Key, StoredValue>>, Self::Error> {
+        let txn = self.environment.create_read_txn()?;
+        let ret = match read_with_proof::<
+            Key,
+            StoredValue,
+            lmdb::RoTransaction,
+            LmdbTrieStore,
+            Self::Error,
+        >(
+            correlation_id,
+            &txn,
+            self.store.deref(),
+            &self.root_hash,
+            key,
+        )? {
+            ReadResult::Found(value) => Some(value),
+            ReadResult::NotFound => None,
+            ReadResult::RootNotFound => panic!("LmdbWithCacheGlobalState has invalid root"),
+        };
+        txn.commit()?;
+        Ok(ret)
+    }
+
+    fn keys_with_prefix(
+        &self,
+        correlation_id: CorrelationId,
+        prefix: &[u8],
+    ) -> Result<Vec<Key>, Self::Error> {
+        let txn = self.environment.create_read_txn()?;
+        let keys_iter = keys_with_prefix::<Key, StoredValue, _, _>(
+            correlation_id,
+            &txn,
+            self.store.deref(),
+            &self.root_hash,
+            prefix,
+        );
+        let mut ret = Vec::new();
+        for result in keys_iter {
+            match result {
+                Ok(key) => ret.push(key),
+                Err(error) => return Err(error),
+            }
+        }
+        txn.commit()?;
+        Ok(ret)
+    }
+}
+
+impl CommitProvider for ScratchGlobalState {
+    /// State hash returned is the one provided, as we do not write to lmdb with this kind of global
+    /// state.
+    fn commit(
+        &self,
+        _correlation_id: CorrelationId,
+        state_hash: Digest,
+        effects: AdditiveMap<Key, Transform>,
+    ) -> Result<Digest, Self::Error> {
+        for (key, transform) in effects.into_iter() {
+            let read_result = self.cache.read().unwrap().get(&key).cloned();
```

Comment:
```suggestion
            let cached_value = self.cache.read().unwrap().get(&key).cloned();
```

b/c it's not `READ` and not really a result.